### PR TITLE
add_2d_array_code_and_ut will add int test next pr

### PR DESCRIPTION
**Pull request by M-iceberg**

### CODE REVIEW

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+
+TEST_FUNCTION_CLEANUP(cleans)
+{
+    umock_c_negative_tests_deinit();
+}
+
+/*TWO_D_ARRAY_CREATE(T)*/
+
+/* Tests_SRS_TWO_D_ARRAY_07_003: [ TWO_D_ARRAY_CREATE(T) shall call THANDLE_MALLOC_FLEX with TWO_D_ARRAY_FREE(T) as dispose function, nmemb set to row_size and size set to sizeof(T*). ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_004: [ TWO_D_ARRAY_CREATE(T) shall set all rows point to NULL. ]*/
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:90]: >rows point [](http://example.com/codeflow?start=71&length=10)

row pointers?

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 2]
```diff
@@ -0,0 +1,642 @@
...
+{
+    //arrange
+    STRICT_EXPECTED_CALL(malloc_flex(IGNORED_ARG, 5, sizeof(uint32_t*)));
+
+    //act
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+    for (unsigned int i = 0; i < 5; i++)
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:102]: >unsigned int [](http://example.com/codeflow?start=9&length=12)

We usually use uint32_t (the fixed size types)

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+/* Tests_SRS_TWO_D_ARRAY_07_003: [ TWO_D_ARRAY_CREATE(T) shall call THANDLE_MALLOC_FLEX with TWO_D_ARRAY_FREE(T) as dispose function, nmemb set to row_size and size set to sizeof(T*). ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_004: [ TWO_D_ARRAY_CREATE(T) shall set all rows point to NULL. ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_006: [ TWO_D_ARRAY_CREATE(T) shall succeed and return a non-NULL value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_CREATE_create_uint32_t_succeeds)
+{
+    //arrange
+    STRICT_EXPECTED_CALL(malloc_flex(IGNORED_ARG, 5, sizeof(uint32_t*)));
+
+    //act
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:98]: >5, 5 [](http://example.com/codeflow?start=63&length=4)

We usually test at the boundaries (inflexion points in the range of values).

Imagine that the range of values for rows is 0 .. UINT32_MAX.
These inflexion points are the points where we expect behavior to be different.

For example I would test what happens when called with 0 - we'd expact that to fail, right?
Also we should test that when called with 1 for rows a 2D array is allocated.

from 1 to UINT32_MAX-1 the expected behavior is the same, so it adds very little value to test for 2 and 3 and 4. So a best practice is to pick something in the interval and test to see the behavior is kept (5 is Ok to test with for example).

So, we should have tests for row values:
0, 1, 5, UINT32_MAX - 1 and UINT32_MAX).

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_003: [ TWO_D_ARRAY_CREATE(T) shall call THANDLE_MALLOC_FLEX with TWO_D_ARRAY_FREE(T) as dispose function, nmemb set to row_size and size set to sizeof(T*). ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_004: [ TWO_D_ARRAY_CREATE(T) shall set all rows point to NULL. ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_006: [ TWO_D_ARRAY_CREATE(T) shall succeed and return a non-NULL value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_CREATE_create_THANDLE_succeeds)
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:134]: >_THANDLE_ [](http://example.com/codeflow?start=39&length=9)

Consider being more explicit in the test name. Like 
TWO_D_ARRAY_CREATE_create_with_THANDLE_type_succeeds

Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:134]: >TWO_D_ARRAY_CREATE_create_THANDLE_succeeds [](http://example.com/codeflow?start=14&length=42)

Nice test

Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:134]: >TWO_D_ARRAY_CREATE_create_THANDLE_succeeds [](http://example.com/codeflow?start=14&length=42)

Do we have a test with a struct type?

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+
+    //act
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
+
+    //assert
+    ASSERT_IS_NULL(tdarr);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+}
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:188]: Consider using umock_negative_tests
We usually use that as a general pattern in all the when_underlying_calls_fails_XXX_api_fails

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5000000);
+    umock_c_reset_all_calls();
+    STRICT_EXPECTED_CALL(malloc_2(5000000, sizeof(uint32_t)));
+
+    //act
+    result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, 0);
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(int, result, 0);
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:236]: This assert should be between ln 227 and 228.
We should not proceed with the test if we were unable to allocate the array in the first place

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_NOT_EQUAL(int, result, 0);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_016: [ If the row specified by row_index has already been allocated, TWO_D_ARRAY_ALLOCATE_NEW_ROW(T) shall fail and return a non-zero value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_ALLOCATE_NEW_ROW_fails_when_that_row_has_already_been_added)
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:338]: >added [](http://example.com/codeflow?start=80&length=5)

I think add is misleading, allocated is the correct term me thinks.
add usually implies that anything existing stays the same and something is added to the collection.

Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:338]: >TWO_D_ARRAY_ALLOCATE_NEW_ROW_fails_when_that_row_has_already_been_added [](http://example.com/codeflow?start=14&length=71)

Is there a test that verifies all rows can be allocated?

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_011: [ If row_index is equal or greater than row_size, TWO_D_ARRAY_FREE_ROW(T) shall fail and return a non-zero value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_FREE_ROW_fails_when_row_index_exceeds)
+{
+    //arrange
+    int result;
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
+    int add_row_result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, 0);
+    ASSERT_IS_NOT_NULL(tdarr);
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:474]: This assert has to be before we try to use the tdarr
Here and everywhere else where applicable.

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+
+    //assert
+    ASSERT_ARE_NOT_EQUAL(int, result, 0);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_011: [ If row_index is equal or greater than row_size, TWO_D_ARRAY_FREE_ROW(T) shall fail and return a non-zero value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_FREE_ROW_fails_when_row_index_exceeds)
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:468]: >exceeds [](http://example.com/codeflow?start=56&length=7)

Consider saying too_high

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+    uint32_t* row_result;
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
+    int add_row_result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, 0);
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(int, add_row_result, 0);
+    ASSERT_IS_NOT_NULL(tdarr->row_arrays[0]);
+    umock_c_reset_all_calls();
+
+    //act
+    row_result = TWO_D_ARRAY_GET(uint32_t)(tdarr, 1);
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:598]: >GET [](http://example.com/codeflow?start=29&length=3)

This is really get_row. We should name it like that

**Review by dcristoloveanu — CHANGES_REQUESTED** [review]
:clock1:

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 3 above)
Comment (dcristoloveanu) [reply tests/two_d_array_ut/two_d_array_ut.c:98]: I guess this would be UINT64_MAX, so you won't actually be able to allocate that much memory, but you can trick the test to not allocate anything, but rather just return a dummy pointer.

Comment (dcristoloveanu) [reply tests/two_d_array_ut/two_d_array_ut.c:98]: The beauty of hooks

File: inc/c_util/two_d_array_ll.h [hunk 11]
```diff
@@ -15,10 +15,236 @@
...
+        /* Codes_SRS_TWO_D_ARRAY_07_003: [ TWO_D_ARRAY_CREATE(T) shall call THANDLE_MALLOC_FLEX with TWO_D_ARRAY_FREE(T) as dispose function, nmemb set to row_size and size set to sizeof(T*). ]*/\
+        result = THANDLE_MALLOC_FLEX(TWO_D_ARRAY_TYPEDEF_NAME(C))(TWO_D_ARRAY_LL_FREE_NAME(C), row_size, sizeof(T*));                                                                              \
+        /* Codes_SRS_TWO_D_ARRAY_07_005: [ If there are any errors then TWO_D_ARRAY_CREATE(T) shall fail and return NULL. ]*/                                                                      \
+        if (result == NULL)                                                                                                                                                                        \
+        {                                                                                                                                                                                          \
+            LogError("failure in malloc_flex(sizeof(TWO_D_ARRAY_HANDLE)=%zu, row_size=%" PRIu64 ", sizeof(unsigned char*)=%zu);", sizeof(T*), row_size, sizeof(unsigned char*));                   \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array_ll.h:118]: > sizeof(unsigned char*) [](http://example.com/codeflow?start=98&length=23)

This should be sizeof(T*)

Comment (dcristoloveanu) [reply inc/c_util/two_d_array_ll.h:118]: And we should print the T with MU_TO_STRING) so that the log line has information about what we're working with.

File: inc/c_util/two_d_array_ll.h [hunk 12]
```diff
@@ -15,10 +15,236 @@
...
+        }                                                                                                                                                           \
+        else                                                                                                                                                        \
+        {                                                                                                                                                           \
+            /* Codes_SRS_TWO_D_ARRAY_07_012: [ TWO_D_ARRAY_FREE_ROW(T) shall free the memory associated with the row specified by row_index and set it to NULL. ]*/ \
+            void* nullptr = NULL;                                                                                                                                   \
+            free((void*)two_d_array->row_arrays[row_index]);                                                                                                        \
+            (void)memcpy((void*)&two_d_array->row_arrays[row_index], &nullptr, sizeof(void*));                                                                      \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array_ll.h:160]: >(void)memcpy((void*)&two_d_array->row_arrays[row_index], &nullptr, sizeof(void*));                                                                      \ [](http://example.com/codeflow?start=12&length=153)

This is an interesting trick. Even for me :-). Why is this needed?

File: inc/c_util/two_d_array_ll.h [hunk 13]
```diff
@@ -15,10 +15,236 @@
...
+                    LogError("malloc(sizeof(col)=%zu) failed", two_d_array->cols * sizeof(T));                                                                                               \
+                    result = MU_FAILURE;                                                                                                                                                     \
+                }                                                                                                                                                                            \
+                else                                                                                                                                                                         \
+                {                                                                                                                                                                            \
+                    (void)memcpy((void*)&two_d_array->row_arrays[row_index], &new_row, sizeof(T*));                                                                                          \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array_ll.h:201]: >(void)memcpy((void*)&two_d_array->row_arrays[row_index], &new_row, sizeof(T*));  [](http://example.com/codeflow?start=20&length=80)

I don't think we need to call memcpy to set the row, do we?

File: inc/c_util/two_d_array_ll.h [hunk 14]
```diff
@@ -15,10 +15,236 @@
...
+            }                                                                                                                                                                                \
+            else                                                                                                                                                                             \
+            {                                                                                                                                                                                \
+                /* Codes_SRS_TWO_D_ARRAY_07_016: [ If the row specified by row_index has already been allocated, TWO_D_ARRAY_ALLOCATE_NEW_ROW(T) shall fail and return a non-zero value. ]*/ \
+                result = MU_FAILURE;                                                                                                                                                         \
+            }                                                                                                                                                                                \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array_ll.h:209]: If this is really an error we should LogError here.
Also we usually put the error first, so consider reversing the logic in 189 to check for != NULL, log the error on the true branch and deal with the allocation on the false branch of the if.

File: inc/c_util/two_d_array_ll.h [hunk 15]
```diff
@@ -15,10 +15,236 @@
...
+            if(result == NULL)                                                                                                                             \
+            {                                                                                                                                              \
+            /* Codes_SRS_TWO_D_ARRAY_07_021: [ If the array stored in row_index is NULL, TWO_D_ARRAY_GET(T) shall fail and return NULL. ]*/                \
+                LogError("The row trying to acquire is not allocated: TWO_D_ARRAY_HANDLE two_d_array=%p, row_index = %" PRIu64, two_d_array, row_index);   \
+            }                                                                                                                                              \
+            /* Codes_SRS_TWO_D_ARRAY_07_022: [ Otherwise, TWO_D_ARRAY_GET(T) shall return the entire column stored in the corresponding row_index. ]*/     \
+            else{                                                                                                                                          \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array_ll.h:240]: >{ [](http://example.com/codeflow?start=16&length=1)

Nitpick:indentation.

File: inc/c_util/two_d_array_ll.h [hunk 16]
```diff
@@ -15,10 +15,236 @@
...
+        }                                                                                                                                                  \
+        else                                                                                                                                               \
+        {                                                                                                                                                  \
+            result = two_d_array->row_arrays[row_index];                                                                                                   \
+            if(result == NULL)                                                                                                                             \
+            {                                                                                                                                              \
+            /* Codes_SRS_TWO_D_ARRAY_07_021: [ If the array stored in row_index is NULL, TWO_D_ARRAY_GET(T) shall fail and return NULL. ]*/                \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array_ll.h:236]: >/*  [](http://example.com/codeflow?start=12&length=3)

Nitpick: add a tab here to indent the Codes like the line of code.

File: inc/c_util/two_d_array_ll.h [hunk 17]
```diff
@@ -15,10 +15,236 @@
...
+T* TWO_D_ARRAY_LL_GET(C)(TWO_D_ARRAY_LL(T) two_d_array, uint64_t row_index)                                                                                \
+{                                                                                                                                                          \
+    T* result;                                                                                                                                             \
+    /* Codes_SRS_TWO_D_ARRAY_07_019: [ If two_d_array is NULL, TWO_D_ARRAY_GET(T) shall fail return NULL. ]*/                                              \
+    if (two_d_array == NULL)                                                                                                                               \
+    {                                                                                                                                                      \
+        LogError("Invalid arguments: TWO_D_ARRAY_HANDLE two_d_array=%p", two_d_array);                                                                     \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array_ll.h:222]: >TWO_D_ARRAY_HANDLE  [](http://example.com/codeflow?start=37&length=19)

All the loglines should printf 
"TWO_D_ARRAY(" MU_TOSTRING(T) ")"

so that we know what we're dealing with as structures rather than a generic TWO_D_ARRAY_HANDLE which does not even exist :-)

**Comment by dcristoloveanu** [comment]
> #include <stdbool.h>

bool is not used, so likely not needed

---
Refers to: inc/c_util/two_d_array_ll.h:10 in ea9bd85. [](commit_id = ea9bd85a5dca269e8d2724629a6790cef1aa6757, deletion_comment = False)

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 18]
```diff
@@ -0,0 +1,642 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+#include <stdlib.h>
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:5]: Pretty sure you want to include inttypes also.

File: inc/c_util/two_d_array.h [hunk 19]
```diff
@@ -10,15 +11,54 @@
 #endif
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array.h:12]: This is not needed in this header, no usage of the PRI macros or the fixed size types.

Comment (M-iceberg) [reply inc/c_util/two_d_array.h:12]: removed

**Comment by M-iceberg** [comment]
> #include <stdbool.h>

removed

---
In reply to: [1931201352](https://github.com/Azure/c-util/pull/352#issuecomment-1931201352) [](http://example.com/codeflow?ancestors=1931201352)

---
Refers to: inc/c_util/two_d_array_ll.h:10 in ea9bd85. [](commit_id = ea9bd85a5dca269e8d2724629a6790cef1aa6757, deletion_comment = False)

File: inc/c_util/two_d_array_ll.h (hunk 15 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array_ll.h:240]: fixed

File: inc/c_util/two_d_array_ll.h (hunk 16 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array_ll.h:236]: fixed

File: inc/c_util/two_d_array_ll.h (hunk 11 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array_ll.h:118]: done

File: inc/c_util/two_d_array_ll.h (hunk 14 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array_ll.h:209]: added the logError and changed the order

File: inc/c_util/two_d_array_ll.h (hunk 17 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array_ll.h:222]: changed to use MU_TOSTRING here and everywhere else

File: inc/c_util/two_d_array_ll.h (hunk 13 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array_ll.h:201]: removed memcpy

File: inc/c_util/two_d_array_ll.h (hunk 12 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array_ll.h:160]: removed memcpy

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 1 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:90]: modified

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 2 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:102]: changed to use uint32_t here and everywhere else

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 18 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:5]: added

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 10 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:598]: changed to get_row here and everywhere else

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 9 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:468]: done

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 5 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:188]: changed to use umock_c_negative_test pattern

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 8 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:474]: modified the order here and everywhere else

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 6 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:236]: moved line 235 to between line 227 and 228

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 4 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:134]: modified

Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:134]: added the test with struct type

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 7 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:338]: TWO_D_ARRAY_ALLOCATE_NEW_ROW_with_uint32_t_type_all_rows_can_be_allocated is testing if all the rows can be allocated in line 310 to line 363

Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:338]: changed to allocated

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 3 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:98]: Thank you so much for explaining the details. I tried to test UINT64_MAX before and that much memory can't be allocated. I try to return a dummy pointer now, but in the two_d_array_ll.h line 125-128, it try to access the returned address by malloc and set every row to NULL, which will give a access violation when I set return a dummy pointer in malloc_flex. I'm confused about whether UINT64_MAX could be tested here.

File: inc/c_util/two_d_array_ll.h [hunk 20]
```diff
@@ -7,18 +7,248 @@
...
+                T* new_row = malloc_2(two_d_array->cols, sizeof(T));                                                                                                                         \
+                if (new_row == NULL)                                                                                                                                                         \
+                {                                                                                                                                                                            \
+                    /* Codes_SRS_TWO_D_ARRAY_07_018: [ If there are any errors then TWO_D_ARRAY_ALLOCATE_NEW_ROW(T) shall fail and return a non-zero value. ]*/                              \
+                    LogError("failure in malloc_2(number of cols=%" PRIu64 ", sizeof(" MU_TOSTRING(T) ")=%zu)",                                                                              \
+                    two_d_array->cols, sizeof(T));                                                                                                                                           \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array_ll.h:202]: >two_d_array->cols, sizeof(T));     [](http://example.com/codeflow?start=20&length=34)

This should be indented, since it belongs to the line above.

File: inc/c_util/two_d_array_ll.h [hunk 21]
```diff
@@ -7,18 +7,248 @@
...
+        {                                                                                                                                                  \
+            result = two_d_array->row_arrays[row_index];                                                                                                   \
+            if(result == NULL)                                                                                                                             \
+            {                                                                                                                                              \
+                /* Codes_SRS_TWO_D_ARRAY_07_021: [ If the array stored in row_index is NULL, TWO_D_ARRAY_GET_ROW(T) shall fail and return NULL. ]*/            \
+                LogError("The row trying to acquire is not allocated: TWO_D_ARRAY (" MU_TOSTRING(T) ") two_d_array=%p, row_index = %" PRIu64,              \
+                two_d_array, row_index);                                                                                                                   \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array_ll.h:240]: >two_d_array, row_index);      [](http://example.com/codeflow?start=16&length=29)

Same here, indentation needed

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 3 above)
Comment (dcristoloveanu) [reply tests/two_d_array_ut/two_d_array_ut.c:131]: Yes, there is no computer with that kind of memory.
For mocking, yes I was thinking that you can return a dummy value, but looks like because the constructor is initializing the memory that won't be doable (without extracting the row access and more mocking it, which would be rather unnecessary).

So we won't write any test with UINT64_MAX I assume.
How about a test with UINT32_MAX + 1, do we test that?

Comment (dcristoloveanu) [reply tests/two_d_array_ut/two_d_array_ut.c:131]: Do we really need uint64_t? I don't think we expect having 4B rows or 4B columns.
Frankly uint32_t should be enough IMO. Then you can also test the extremeties of the interval.

File: inc/c_util/two_d_array.h [hunk 22]
```diff
@@ -1,24 +1,58 @@
...
 
-/*macro to be used in headers*/                                                                                     \
+/*macro to be used in headers*/                                                                                       \
 #define TWO_D_ARRAY_TYPE_DECLARE(T)                                                                                   \
+    /*hint: have TWO_D_ARRAY_TYPE_DEFINE(T) before TWO_D_ARRAY_TYPE_DECLARE                                        */ \
+    /*hint: have THANDLE_TYPE_DECLARE(TWO_D_ARRAY_TYPEDEF_NAME(T)) before TWO_D_ARRAY_TYPE_DECLARE                 */ \
+    TWO_D_ARRAY_CREATE_DECLARE(T)                                                                                     \
+    TWO_D_ARRAY_FREE_ROW_DECLARE(T)                                                                                   \
+    TWO_D_ARRAY_ALLOCATE_NEW_ROW_DECLARE(T)                                                                           \
+    TWO_D_ARRAY_GET_ROW_DECLARE(T)                                                                                        \
```
Comment (dcristoloveanu) [inline inc/c_util/two_d_array.h:49]: Let's tidy up the \ to have them all nicely aligned

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 3 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:131]: done

File: inc/c_util/two_d_array_ll.h (hunk 21 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array_ll.h:240]: done

File: inc/c_util/two_d_array_ll.h (hunk 20 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array_ll.h:202]: done

File: inc/c_util/two_d_array.h (hunk 22 above)
Comment (M-iceberg) [reply inc/c_util/two_d_array.h:49]: done

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 23]
```diff
@@ -0,0 +1,811 @@
...
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+}
+*/
+
+/* Tests_SRS_TWO_D_ARRAY_07_003: [ TWO_D_ARRAY_CREATE(T) shall call THANDLE_MALLOC_FLEX with TWO_D_ARRAY_FREE(T) as dispose function, nmemb set to row_size and size set to sizeof(T*). ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_004: [ TWO_D_ARRAY_CREATE(T) shall set all rows pointers to NULL. ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_006: [ TWO_D_ARRAY_CREATE(T) shall succeed and return a non-NULL value. ]*/
+/* running too long in the gate, comment out
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:169]: >comment [](http://example.com/codeflow?start=33&length=7)

Nitpick:commented

File: tests/two_d_array_ut/two_d_array_ut.c [hunk 24]
```diff
@@ -0,0 +1,811 @@
...
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_003: [ TWO_D_ARRAY_CREATE(T) shall call THANDLE_MALLOC_FLEX with TWO_D_ARRAY_FREE(T) as dispose function, nmemb set to row_size and size set to sizeof(T*). ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_004: [ TWO_D_ARRAY_CREATE(T) shall set all rows pointers to NULL. ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_006: [ TWO_D_ARRAY_CREATE(T) shall succeed and return a non-NULL value. ]*/
+/* running too long in the gate, comment out
```
Comment (dcristoloveanu) [inline tests/two_d_array_ut/two_d_array_ut.c:148]: >/* [](http://example.com/codeflow?start=0&length=2)

Can we have the tests disabled with #if 0 and #endif
It makes it much easier to enable if someone wants to run the tests - whoever wants to run them would change it to be #if 1.

**Review by dcristoloveanu — DISMISSED** [review]
:shipit:

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 24 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:148]: done

File: tests/two_d_array_ut/two_d_array_ut.c (hunk 23 above)
Comment (M-iceberg) [reply tests/two_d_array_ut/two_d_array_ut.c:169]: done

**Review by dcristoloveanu — APPROVED** [review]
:shipit:
